src/wallet/wallet.cpp

```diff
@@ -1896,177 +2341,316 @@ bool CWalletTx::RelayWalletTransaction()
-        return 0;
+    int nDepth = GetDepthInMainChain();
+    if (nDepth < 0)
+        return CCoinsViewCache::outputMaturity::NOT_APPLICABLE;
+
+    if (!IsCoinBase() && !IsCertificate())
+        return CCoinsViewCache::outputMaturity::MATURE;
+
+    if (IsCoinBase())
+    {
+        if (nDepth > COINBASE_MATURITY)
+            return CCoinsViewCache::outputMaturity::MATURE;
+        else
+            return CCoinsViewCache::outputMaturity::IMMATURE;
+    }
+
+    if (IsCertificate())
+    {
+        // we do not consider even an unconfirmed change output in a certificate as mature
+        if ((nDepth == 0))
+            return CCoinsViewCache::outputMaturity::IMMATURE;
+
+        return pcoinsTip->IsCertOutputMature(hash, vOutPos, chainActive.Height());
+    }
+
+    return CCoinsViewCache::outputMaturity::NOT_APPLICABLE;
+}
+
+CAmount CWalletObjBase::GetCredit(const isminefilter& filter) const
+{
+    int64_t credit = 0;
+    if (filter & ISMINE_SPENDABLE) {
+        // It used to be that GetBalance can assume transactions in mapWallet won't change
+        // With certificate it is up to the transaction to tell whether its credit
+        // won't change anymore
+        if (!fCreditCached)
+            nCreditCached = pwallet->GetCredit(*this, ISMINE_SPENDABLE, fCreditCached, /*keepImmatureVoutsOnly*/false);
+
+        credit += nCreditCached;
+    }
 
-    // Must wait until coinbase is safely deep enough in the chain before valuing it
-    if (IsCoinBase() && GetBlocksToMaturity() > 0)
+    if (filter & ISMINE_WATCH_ONLY) {
+        // Again here, with certificate it is up to the transaction to tell whether its credit
+        // won't change anymore
+        if (!fWatchCreditCached)
+            nWatchCreditCached = pwallet->GetCredit(*this, ISMINE_WATCH_ONLY, fWatchCreditCached, /*keepImmatureVoutsOnly*/false);
+
+        credit += nWatchCreditCached;
+    }
+
+    return credit;
+}
+
+CAmount CWalletObjBase::GetImmatureCredit(bool fUseCache) const
+{
+    if (!IsInMainChain() && !IsCertificate())
+        return CAmount(0);
+
+    if (!IsCoinBase() && !IsCertificate())
+        return CAmount(0);
+
+    if (fUseCache && fImmatureCreditCached)
+        return nImmatureCreditCached;
+
+    nImmatureCreditCached = pwallet->GetCredit(*this, ISMINE_SPENDABLE, fImmatureCreditCached, /*keepImmatureVoutsOnly*/true);
+    return nImmatureCreditCached;
+
+}
+
+CAmount CWalletObjBase::GetImmatureWatchOnlyCredit(const bool& fUseCache) const
+{
+    if (!IsInMainChain())
+        return CAmount(0);
+
+    if (!IsCoinBase() && !IsCertificate())
+        return CAmount(0);
+
+    if (fUseCache && fImmatureWatchCreditCached)
+        return nImmatureWatchCreditCached;
+
+    nImmatureWatchCreditCached = pwallet->GetCredit(*this, ISMINE_WATCH_ONLY, fImmatureWatchCreditCached, /*keepImmatureVoutsOnly*/true);
+    return nImmatureWatchCreditCached;
+}
+
+CAmount CWalletObjBase::GetAvailableCredit(bool fUseCache) const
+{
+    if (pwallet == 0)
```

Comment:
we should not return the cached value if one of the outputs became immature (like in the original version). This can happen for example in case of a cert became mature and then block revert happened. Moreover it can happen that a cert became mature at the boundary of the acceptance window, the a block revert happened and a new block arrived without a cert. This will cause the sidechain to cease, then previous certificate bwt are "voided" but cached credit would return also the amount of voided bwts. This last case would be solved after the introduction of the ceased sidechain event.
In any case we need to have an efficient way to detect non mature bwts  because it will be made everytime before using the cache. This will be possible by storing the needed info as described above.